// @redbubble/design-system/react/headerAndFooter/components/HeaderLinks/index.js
import React from 'react';
import PropTypes from 'prop-types';
import { defineMessages } from 'react-intl';
import Button from '../../../Button';

const messages = defineMessages({
  sellArt: 'Sell your art',
  login: 'Login',
  signup: 'Signup',
});

const HeaderLinks = ({
  intl,
  baseUrl,
  doLogin,
  doSignup,
  isDesktop,
  isLoggedIn,
}) => {
  if (!isDesktop || isLoggedIn) return null;

  const baseLinkProps = {
    rel: 'nofollow',
  };

  const loginProps = {
    ...baseLinkProps,
    href: `${baseUrl}/auth/login`,
  };
  if (doLogin) loginProps.onClick = doLogin;

  const signupProps = {
    ...baseLinkProps,
    href: `${baseUrl}/signup`,
  };
  if (doSignup) signupProps.onClick = doSignup;

  return (
    <>
      <Button href={`${baseUrl}/about/selling`}>
        { intl.formatMessage(messages.sellArt) }
      </Button>
      <Button
        {...loginProps}
        data-testid="ds-header-login-action"
      >
        { intl.formatMessage(messages.login) }
      </Button>
      <Button
        {...signupProps}
        data-testid="ds-header-signup-action"
      >
        { intl.formatMessage(messages.signup) }
      </Button>
    </>
  );
};

HeaderLinks.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  baseUrl: PropTypes.string.isRequired,
  doLogin: PropTypes.func,
  doSignup: PropTypes.func,
};


HeaderLinks.defaultProps = {
  doLogin: null,
  doSignup: null,
};

HeaderLinks.displayName = 'HeaderLinks';

export default HeaderLinks;
