// @redbubble/boom-analytics/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.middleware = exports.reducer = exports.actions = exports.constants = exports.withAnalytics = exports.DataLayer = exports.createGaClient = undefined;

var _DataLayer = require('./DataLayer');

Object.defineProperty(exports, 'DataLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_DataLayer).default;
  }
});

var _withAnalytics = require('./react/withAnalytics');

Object.defineProperty(exports, 'withAnalytics', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_withAnalytics).default;
  }
});

var _reactGa2 = require('react-ga');

var _reactGa3 = _interopRequireDefault(_reactGa2);

var _constants = require('./constants');

var constants = _interopRequireWildcard(_constants);

var _actions = require('./redux/actions');

var actions = _interopRequireWildcard(_actions);

var _reducer = require('./redux/reducer');

var _reducer2 = _interopRequireDefault(_reducer);

var _middleware = require('./redux/middleware');

var _middleware2 = _interopRequireDefault(_middleware);

var _GaClient = require('./GaClient');

var _GaClient2 = _interopRequireDefault(_GaClient);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createGaClient = exports.createGaClient = function createGaClient(account) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  if (typeof window === 'undefined') {
    throw new Error('createGaClient should only be called client side.');
  }

  return new _GaClient2.default(account, _reactGa3.default, window, options);
};

exports.constants = constants;
exports.actions = actions;
exports.reducer = _reducer2.default;
exports.middleware = _middleware2.default;
