// @redbubble/design-system/react/headerAndFooter/components/Footer/utils.js

import { defineMessages } from 'react-intl';

export const AMAZONPAY = 'amazonpay';
export const AMERICAN_EXPRESS = 'american_express';
export const APPLEPAY = 'applepay';
export const MASTERCARD = 'mastercard';
export const PAYPAL = 'paypal';
export const SEPA = 'sepa';
export const SOFORT = 'sofort';
export const VISA = 'visa';

const messages = defineMessages({
  visa: 'Visa',
  masterCard: 'MasterCard',
  americanExpress: 'American Express',
  payPal: 'PayPal',
  applePay: 'Apple pay',
  amazonPay: 'Amazon Pay',
  sofort: 'Sofort',
});

export const availablePaymentMethods = (intl, currency = '') => {
  const universalPaymentMethods = [{
    icon: VISA,
    text: intl.formatMessage(messages.visa),
  }, {
    icon: MASTERCARD,
    text: intl.formatMessage(messages.masterCard),
  }, {
    icon: AMERICAN_EXPRESS,
    text: intl.formatMessage(messages.americanExpress),
  }, {
    icon: PAYPAL,
    text: intl.formatMessage(messages.payPal),
  }, {
    icon: APPLEPAY,
    text: intl.formatMessage(messages.applePay),
  }];

  if (currency === 'USD') {
    return universalPaymentMethods.concat([{
      icon: AMAZONPAY,
      text: intl.formatMessage(messages.amazonPay),
    }]);
  }

  if (currency === 'EUR') {
    return universalPaymentMethods.concat([{
      icon: SOFORT,
      text: intl.formatMessage(messages.sofort),
    }]);
  }

  return universalPaymentMethods;
};
